File: engine/src/theseus_engine/dstar_lite.py
"""D* Lite — incremental replanning (Koenig & Likhachev, AAAI 2002).

Why this algorithm is the heart of Theseus: the world model changes every
few sensor ticks (a person walks through the corridor, a door opens, a
chair moves). Replanning from scratch with A* on every change wastes work;
D* Lite repairs only the part of the search affected by the change, and it
searches BACKWARD from the goal so that the agent moving (= the start
changing) is also cheap, handled by the km "key modifier" trick instead of
recomputing heuristics for the whole queue.

Invariants worth knowing when reading this file:
  g(u)   = best known cost from u to the GOAL (backward search)
  rhs(u) = one-step lookahead of g(u): min over successors s of
           c(u, s) + g(s); rhs(goal) = 0
  A vertex is "consistent" when g == rhs. The queue holds exactly the
  inconsistent vertices, ordered by calculate_key(). When
  compute_shortest_path() finishes, following the greedy successor
  argmin(c(u,s) + g(s)) from the start yields an optimal path.

This implementation recomputes rhs values exactly (a full min over
successors) instead of the paper's incremental shortcuts. It costs a few
extra neighbor evaluations per touched vertex and in exchange removes the
subtlest class of D* Lite bugs. The A*-equivalence property tests in
tests/test_dstar_lite.py are the safety net for the Swift port.

The priority queue uses lazy deletion: `entry[u]` remembers the sequence
number of u's latest valid heap entry; anything else popped is stale and
skipped.
"""

from __future__ import annotations

from heapq import heappop, heappush

from .geometry import Cell, INF, octile
from .grid import OccupancyGrid, PlanParams
from .astar import PlanResult

_EPS = 1e-9

Key = tuple[float, float]


class DStarLite:
    def __init__(self, grid: OccupancyGrid, start: Cell, goal: Cell,
                 params: PlanParams):
        self.grid = grid
        self.params = params
        self.start = start
        self.goal = goal
        self.km = 0.0
        self.s_last = start
        self.g: dict[Cell, float] = {}
        self.rhs: dict[Cell, float] = {goal: 0.0}
        self._heap: list[tuple[float, float, int, Cell]] = []
        self._entry: dict[Cell, int | None] = {}
        self._seq = 0
        self._push(goal)

    # ---- small accessors ----------------------------------------------

    def _gv(self, u: Cell) -> float:
        return self.g.get(u, INF)

    def _rhsv(self, u: Cell) -> float:
        return self.rhs.get(u, INF)

    def _h(self, u: Cell) -> float:
        return octile(self.start, u) * self.grid.cell_size

    def _key(self, u: Cell) -> Key:
        v = min(self._gv(u), self._rhsv(u))
        if v == INF:
            return (INF, INF)
        return (v + self._h(u) + self.km, v)

    # ---- lazy-deletion priority queue -----------------------------------

    def _push(self, u: Cell) -> None:
        k = self._key(u)
        self._seq += 1
        self._entry[u] = self._seq
        heappush(self._heap, (k[0], k[1], self._seq, u))

    def _invalidate(self, u: Cell) -> None:
        self._entry[u] = None

    def _top_key(self) -> Key:
        while self._heap and self._entry.get(self._heap[0][3]) != self._heap[0][2]:
            heappop(self._heap)
        if not self._heap:
            return (INF, INF)
        return (self._heap[0][0], self._heap[0][1])

    def _pop(self) -> tuple[Key, Cell] | None:
        while self._heap:
            k1, k2, seq, u = heappop(self._heap)
            if self._entry.get(u) == seq:
                self._entry[u] = None
                return (k1, k2), u
        return None

    def _update_vertex(self, u: Cell) -> None:
        if abs(self._gv(u) - self._rhsv(u)) > _EPS:
            self._push(u)
        else:
            self._invalidate(u)

    def _recompute_rhs(self, u: Cell) -> None:
        if u == self.goal:
            self._update_vertex(u)
            return
        best = INF
        for v, _step in self.grid.neighbors8(u):
            c = self.grid.edge_cost(u, v, self.params)
            if c == INF:
                continue
            cand = c + self._gv(v)
            if cand < best:
                best = cand
        self.rhs[u] = best
        self._update_vertex(u)

    # ---- core -----------------------------------------------------------

    def _compute_shortest_path(self) -> None:
        guard = 0
        limit = 40 * self.grid.width * self.grid.height  # safety net only
        while True:
            guard += 1
            if guard > limit:
                raise RuntimeError("D* Lite failed to converge (bug)")
            top = self._top_key()
            ks = self._key(self.start)
            start_consistent = abs(self._rhsv(self.start) - self._gv(self.start)) <= _EPS
            if not (top < ks) and start_consistent:
                break
            popped = self._pop()
            if popped is None:
                break
            k_old, u = popped
            k_new = self._key(u)
            if k_old < k_new:
                # key went stale (km grew or costs changed): reorder, don't process
                self._push(u)
                continue
            gu, ru = self._gv(u), self._rhsv(u)
            if gu > ru + _EPS:
                # overconsistent: this g value is now final
                self.g[u] = ru
                for s, _step in self.grid.neighbors8(u):
                    self._recompute_rhs(s)
            else:
                # underconsistent: invalidate and let the wave repair it
                self.g[u] = INF
                self._recompute_rhs(u)
                for s, _step in self.grid.neighbors8(u):
                    self._recompute_rhs(s)

    def plan(self) -> PlanResult | None:
        """(Re)compute and extract the current optimal path start -> goal."""
        self._compute_shortest_path()
        if self._gv(self.start) == INF:
            return None
        cells = [self.start]
        cost = 0.0
        u = self.start
        cap = self.grid.width * self.grid.height + 1
        while u != self.goal:
            best_v: Cell | None = None
            best = INF
            for v, _step in self.grid.neighbors8(u):
                c = self.grid.edge_cost(u, v, self.params)
                if c == INF:
                    continue
                cand = c + self._gv(v)
                if cand < best - _EPS:
                    best = cand
                    best_v = v
                    best_c = c
            if best_v is None or best == INF:
                return None
            cost += best_c
            u = best_v
            cells.append(u)
            if len(cells) > cap:
                return None  # should be unreachable once converged
        return PlanResult(cells, cost)

    # ---- world interface --------------------------------------------------

    def update_start(self, new_start: Cell) -> None:
        """Call whenever the agent's cell changes. km absorbs the heuristic
        drift so stale queue keys stay comparable (the D* Lite trick)."""
        if new_start == self.start:
            return
        self.km += self._h(new_start)  # h is measured from current start: octile(start, new)
        self.start = new_start
        self.s_last = new_start

    def notify_changed(self, cells) -> None:
        """Call with every cell whose derived state changed. Edge costs
        depend only on the two endpoint cells plus (for diagonals) their
        shared cardinal neighbors, all of which lie within distance 1 of a
        changed cell — so recomputing rhs for the changed cells and their
        8-neighborhoods repairs every affected edge."""
        affected: set[Cell] = set()
        for c in cells:
            c = tuple(c)
            if not self.grid.in_bounds(c):
                continue
            affected.add(c)
            for v, _step in self.grid.neighbors8(c):
                affected.add(v)
        for u in affected:
            self._recompute_rhs(u)

File: apps/navcore/Sources/NavCore/DStarLite.swift
// D* Lite — incremental replanning (Koenig & Likhachev, AAAI 2002).
// Port of engine dstar_lite.py; read that file's docstring for the full
// account of the invariants (g / rhs / consistency / the km trick).
//
// Like the Python, this recomputes rhs values exactly (a full min over
// successors) instead of the paper's incremental shortcuts — a few extra
// neighbor evaluations per touched vertex in exchange for removing the
// subtlest class of D* Lite bugs. The A*-equivalence tests are the
// safety net for this port, same as they were for the original.
//
// One deliberate deviation: notifyChanged iterates its affected set in
// sorted order. Python iterates a hash set (arbitrary order); the final
// g/rhs values are order-independent — rhs reads only g, which does not
// change during the sweep — so sorting changes nothing semantically and
// buys run-to-run determinism Swift's randomized Set ordering would lose.

private let EPS = 1e-9

private struct DEntry: Comparable {
    var k1: Double
    var k2: Double
    var seq: Int
    var cell: Cell

    static func < (a: DEntry, b: DEntry) -> Bool {
        if a.k1 != b.k1 { return a.k1 < b.k1 }
        if a.k2 != b.k2 { return a.k2 < b.k2 }
        return a.seq < b.seq
    }

    static func == (a: DEntry, b: DEntry) -> Bool {
        a.k1 == b.k1 && a.k2 == b.k2 && a.seq == b.seq
    }
}

private func keyLess(_ a: (Double, Double), _ b: (Double, Double)) -> Bool {
    if a.0 != b.0 { return a.0 < b.0 }
    return a.1 < b.1
}

public final class DStarLite {
    let grid: OccupancyGrid
    let params: PlanParams
    public private(set) var start: Cell
    public let goal: Cell
    var km = 0.0
    var sLast: Cell
    var g: [Cell: Double] = [:]
    var rhs: [Cell: Double] = [:]
    private var heap = BinaryHeap<DEntry>()
    private var entry: [Cell: Int] = [:]
    private var seq = 0

    public init(grid: OccupancyGrid, start: Cell, goal: Cell,
                params: PlanParams) {
        self.grid = grid
        self.params = params
        self.start = start
        self.goal = goal
        self.sLast = start
        self.rhs[goal] = 0.0
        push(goal)
    }

    // ---- small accessors ----------------------------------------------

    private func gv(_ u: Cell) -> Double { g[u] ?? INF }
    private func rhsv(_ u: Cell) -> Double { rhs[u] ?? INF }

    private func h(_ u: Cell) -> Double {
        octile(start, u) * grid.cellSize
    }

    private func key(_ u: Cell) -> (Double, Double) {
        let v = Swift.min(gv(u), rhsv(u))
        if v == INF {
            return (INF, INF)
        }
        return (v + h(u) + km, v)
    }

    // ---- lazy-deletion priority queue -----------------------------------

    private func push(_ u: Cell) {
        let k = key(u)
        seq += 1
        entry[u] = seq
        heap.push(DEntry(k1: k.0, k2: k.1, seq: seq, cell: u))
    }

    private func invalidate(_ u: Cell) {
        entry[u] = nil
    }

    private func topKey() -> (Double, Double) {
        while let top = heap.min, entry[top.cell] != top.seq {
            _ = heap.pop()
        }
        guard let top = heap.min else {
            return (INF, INF)
        }
        return (top.k1, top.k2)
    }

    private func pop() -> ((Double, Double), Cell)? {
        while let e = heap.pop() {
            if entry[e.cell] == e.seq {
                entry[e.cell] = nil
                return ((e.k1, e.k2), e.cell)
            }
        }
        return nil
    }

    private func updateVertex(_ u: Cell) {
        if abs(gv(u) - rhsv(u)) > EPS {
            push(u)
        } else {
            invalidate(u)
        }
    }

    private func recomputeRhs(_ u: Cell) {
        if u == goal {
            updateVertex(u)
            return
        }
        var best = INF
        for (v, _) in grid.neighbors8(u) {
            let c = grid.edgeCost(u, v, params)
            if c == INF {
                continue
            }
            let cand = c + gv(v)
            if cand < best {
                best = cand
            }
        }
        rhs[u] = best
        updateVertex(u)
    }

    // ---- core -----------------------------------------------------------

    private func computeShortestPath() {
        var guardCount = 0
        let limit = 40 * grid.width * grid.height  // safety net only
        while true {
            guardCount += 1
            precondition(guardCount <= limit,
                         "D* Lite failed to converge (bug)")
            let top = topKey()
            let ks = key(start)
            let startConsistent = abs(rhsv(start) - gv(start)) <= EPS
            if !keyLess(top, ks) && startConsistent {
                break
            }
            guard let (kOld, u) = pop() else {
                break
            }
            let kNew = key(u)
            if keyLess(kOld, kNew) {
                // key went stale (km grew or costs changed): reorder
                push(u)
                continue
            }
            let gu = gv(u)
            let ru = rhsv(u)
            if gu > ru + EPS {
                // overconsistent: this g value is now final
                g[u] = ru
                for (s, _) in grid.neighbors8(u) {
                    recomputeRhs(s)
                }
            } else {
                // underconsistent: invalidate and let the wave repair it
                g[u] = INF
                recomputeRhs(u)
                for (s, _) in grid.neighbors8(u) {
                    recomputeRhs(s)
                }
            }
        }
    }

    /// (Re)compute and extract the current optimal path start -> goal.
    public func plan() -> PlanResult? {
        computeShortestPath()
        if gv(start) == INF {
            return nil
        }
        var cells = [start]
        var cost = 0.0
        var u = start
        let cap = grid.width * grid.height + 1
        while u != goal {
            var bestV: Cell? = nil
            var best = INF
            var bestC = 0.0
            for (v, _) in grid.neighbors8(u) {
                let c = grid.edgeCost(u, v, params)
                if c == INF {
                    continue
                }
                let cand = c + gv(v)
                if cand < best - EPS {
                    best = cand
                    bestV = v
                    bestC = c
                }
            }
            guard let next = bestV, best < INF else {
                return nil
            }
            cost += bestC
            u = next
            cells.append(u)
            if cells.count > cap {
                return nil  // should be unreachable once converged
            }
        }
        return PlanResult(cells: cells, cost: cost)
    }

    // ---- world interface --------------------------------------------------

    /// Call whenever the agent's cell changes. km absorbs the heuristic
    /// drift so stale queue keys stay comparable (the D* Lite trick).
    public func updateStart(_ newStart: Cell) {
        if newStart == start {
            return
        }
        km += h(newStart)  // h measured from the CURRENT start
        start = newStart
        sLast = newStart
    }

    /// Call with every cell whose derived state changed. Edge costs depend
    /// only on the two endpoint cells plus (for diagonals) their shared
    /// cardinal neighbors, all within distance 1 of a changed cell — so
    /// recomputing rhs over the changed cells' 8-neighborhoods repairs
    /// every affected edge.
    public func notifyChanged(_ cells: [Cell]) {
        var affected = Set<Cell>()
        for c in cells {
            if !grid.inBounds(c) {
                continue
            }
            affected.insert(c)
            for (v, _) in grid.neighbors8(c) {
                affected.insert(v)
            }
        }
        for u in affected.sorted(by: { ($0.x, $0.y) < ($1.x, $1.y) }) {
            recomputeRhs(u)
        }
    }
}
